core/src/main/java/kafka/log/remote/RemoteLogManager.java

```diff
@@ -1209,60 +1195,80 @@ void cleanupExpiredRemoteLogSegments() throws RemoteStorageException, ExecutionE
-            Optional<RetentionTimeData> retentionTimeData = buildRetentionTimeData(log.config().retentionMs);
-
-            RemoteLogRetentionHandler remoteLogRetentionHandler = new RemoteLogRetentionHandler(retentionSizeData, retentionTimeData);
-            Iterator<Integer> epochIterator = epochWithOffsets.navigableKeySet().iterator();
-            boolean canProcess = true;
-            List<RemoteLogSegmentMetadata> segmentsToDelete = new ArrayList<>();
             long sizeOfDeletableSegmentsBytes = 0L;
-            while (canProcess && epochIterator.hasNext()) {
-                Integer epoch = epochIterator.next();
-                Iterator<RemoteLogSegmentMetadata> segmentsIterator = remoteLogMetadataManager.listRemoteLogSegments(topicIdPartition, epoch);
-                while (canProcess && segmentsIterator.hasNext()) {
-                    if (isCancelled()) {
-                        logger.info("Returning from remote log segments cleanup for the remaining segments as the task state is changed.");
-                        return;
-                    }
-                    RemoteLogSegmentMetadata metadata = segmentsIterator.next();
 
-                    if (segmentIdsBeingCopied.contains(metadata.remoteLogSegmentId())) {
-                        logger.debug("Copy for the segment {} is currently in process. Skipping cleanup for it and the remaining segments",
-                                metadata.remoteLogSegmentId());
-                        canProcess = false;
-                        continue;
-                    }
-                    if (RemoteLogSegmentState.DELETE_SEGMENT_FINISHED.equals(metadata.state())) {
+            final List<RemoteLogSegmentMetadata> segmentsToDelete = new ArrayList<>();
+            final List<RemoteLogSegmentMetadata> validSegments = new ArrayList<>();
+            for (Integer remoteLeaderEpoch: epochWithOffsets.navigableKeySet()) {
+                Iterator<RemoteLogSegmentMetadata> it = remoteLogMetadataManager.listRemoteLogSegments(topicIdPartition, remoteLeaderEpoch);
+                while (it.hasNext()) {
+                    final RemoteLogSegmentMetadata segment = it.next();
+
+                    // We can remove all segments in COPY_SEGMENT_STARTED but the last one as they are dangling
+                    if (segment.state().equals(RemoteLogSegmentState.COPY_SEGMENT_STARTED) && it.hasNext()) {
+                        sizeOfDeletableSegmentsBytes += segment.segmentSizeInBytes();
+                        segmentsToDelete.add(segment);
                         continue;
                     }
-                    if (segmentsToDelete.contains(metadata)) {
+
+                    if (segment.state().equals(RemoteLogSegmentState.DELETE_SEGMENT_STARTED)) {
+                        segmentsToDelete.add(segment);
+                        sizeOfDeletableSegmentsBytes += segment.segmentSizeInBytes();
                         continue;
                     }
```

Comment:
This is the main fix, right?
Before this fix, we will add all `COPY_SEGMENT_STARTED` segments into validSegments, including retried ones. But here we will filter them out, right?

Reply:
We will check `COPY_SEGMENT_STARTED` and `DELETE_SEGMENT_STARTED` here, but `DELETE_SEGMENT_FINISHED` in L1244. I don't think it's readable. Could you make it clear?